File: backend/app/models/document.py
from datetime import datetime
from enum import StrEnum
from uuid import UUID, uuid4

from sqlalchemy import (
    BigInteger,
    Boolean,
    DateTime,
    Enum,
    Float,
    ForeignKey,
    Integer,
    String,
    Text,
    UniqueConstraint,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class DocumentStatus(StrEnum):
    UPLOADED = "uploaded"
    PROCESSING = "processing"
    READY = "ready"
    FAILED = "failed"


class DocumentPageKind(StrEnum):
    TEXTUAL = "textual"
    SCANNED = "scanned"
    MIXED = "mixed"
    EMPTY = "empty"


class TextExtractionMethod(StrEnum):
    NATIVE = "native"
    OCR = "ocr"
    NONE = "none"


class OcrStatus(StrEnum):
    NOT_REQUIRED = "not_required"
    REQUIRED = "required"
    COMPLETED = "completed"
    FAILED = "failed"


class ChapterDetectionMethod(StrEnum):
    PDF_TOC = "pdf_toc"
    AUTOMATIC_HEADING = "automatic_heading"
    MANUAL = "manual"


class Document(Base):
    __tablename__ = "documents"

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    organization_id: Mapped[UUID] = mapped_column(
        ForeignKey("organizations.id", ondelete="CASCADE"),
        index=True,
    )
    uploaded_by_id: Mapped[UUID] = mapped_column(
        ForeignKey("users.id", ondelete="RESTRICT"),
        index=True,
    )
    project_id: Mapped[UUID | None] = mapped_column(
        ForeignKey("projects.id", ondelete="SET NULL"),
        index=True,
        nullable=True,
    )
    original_filename: Mapped[str] = mapped_column(String(255), nullable=False)
    storage_key: Mapped[str] = mapped_column(String(500), unique=True, nullable=False)
    mime_type: Mapped[str] = mapped_column(String(120), nullable=False)
    size_bytes: Mapped[int] = mapped_column(BigInteger, nullable=False)
    checksum_sha256: Mapped[str] = mapped_column(String(64), index=True, nullable=False)
    status: Mapped[DocumentStatus] = mapped_column(
        Enum(DocumentStatus, name="document_status"),
        default=DocumentStatus.UPLOADED,
        nullable=False,
    )
    page_count: Mapped[int | None] = mapped_column(Integer, nullable=True)
    extracted_text: Mapped[str | None] = mapped_column(Text(), nullable=True)
    extraction_error: Mapped[str | None] = mapped_column(Text(), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )
    processed_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )


class DocumentPage(Base):
    __tablename__ = "document_pages"
    __table_args__ = (
        UniqueConstraint("document_id", "page_number", name="uq_document_page_number"),
    )

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    document_id: Mapped[UUID] = mapped_column(
        ForeignKey("documents.id", ondelete="CASCADE"),
        index=True,
    )
    page_number: Mapped[int] = mapped_column(Integer, nullable=False)
    text: Mapped[str] = mapped_column(Text(), default="", nullable=False)
    character_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    image_count: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    page_kind: Mapped[DocumentPageKind] = mapped_column(
        Enum(DocumentPageKind, name="document_page_kind"),
        nullable=False,
    )
    extraction_method: Mapped[TextExtractionMethod] = mapped_column(
        Enum(TextExtractionMethod, name="text_extraction_method"),
        nullable=False,
    )
    ocr_status: Mapped[OcrStatus] = mapped_column(
        Enum(OcrStatus, name="ocr_status"),
        nullable=False,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )


class DocumentChapter(Base):
    __tablename__ = "document_chapters"

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    document_id: Mapped[UUID] = mapped_column(
        ForeignKey("documents.id", ondelete="CASCADE"),
        index=True,
    )
    title: Mapped[str] = mapped_column(String(240), nullable=False)
    chapter_number: Mapped[int | None] = mapped_column(Integer, nullable=True)
    start_page: Mapped[int] = mapped_column(Integer, nullable=False)
    end_page: Mapped[int] = mapped_column(Integer, nullable=False)
    summary: Mapped[str | None] = mapped_column(Text(), nullable=True)
    detection_method: Mapped[ChapterDetectionMethod] = mapped_column(
        Enum(ChapterDetectionMethod, name="chapter_detection_method"),
        nullable=False,
    )
    confidence: Mapped[float] = mapped_column(Float, default=1.0, nullable=False)
    is_confirmed: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    position: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

File: frontend/src/types/document.ts
export type DocumentStatus =
  | 'uploaded'
  | 'processing'
  | 'ready'
  | 'failed'

export type DocumentPageKind = 'textual' | 'scanned' | 'mixed' | 'empty'
export type TextExtractionMethod = 'native' | 'ocr' | 'none'
export type OcrStatus = 'not_required' | 'required' | 'completed' | 'failed'
export type ChapterDetectionMethod =
  | 'pdf_toc'
  | 'automatic_heading'
  | 'manual'

export interface DocumentItem {
  id: string
  organization_id: string
  uploaded_by_id: string
  project_id?: string | null
  original_filename: string
  mime_type: string
  size_bytes: number
  checksum_sha256: string
  status: DocumentStatus
  page_count?: number | null
  created_at: string
  updated_at: string
  processed_at?: string | null
}

export interface DocumentDetail extends DocumentItem {
  extracted_text?: string | null
  extraction_error?: string | null
}

export interface DocumentPageItem {
  id: string
  document_id: string
  page_number: number
  character_count: number
  image_count: number
  page_kind: DocumentPageKind
  extraction_method: TextExtractionMethod
  ocr_status: OcrStatus
  text_preview: string
}

export interface DocumentPageDetail {
  id: string
  document_id: string
  page_number: number
  text: string
  character_count: number
  image_count: number
  page_kind: DocumentPageKind
  extraction_method: TextExtractionMethod
  ocr_status: OcrStatus
  created_at: string
  updated_at: string
}

export interface DocumentChapter {
  id: string
  document_id: string
  title: string
  chapter_number?: number | null
  start_page: number
  end_page: number
  summary?: string | null
  detection_method: ChapterDetectionMethod
  confidence: number
  is_confirmed: boolean
  position: number
  created_at: string
  updated_at: string
}

export interface ChapterTextPreview {
  chapter: DocumentChapter
  text: string
  character_count: number
  source_pages: number[]
}

export interface DocumentStructureSummary {
  document_id: string
  page_count: number
  extracted_pages: number
  textual_pages: number
  mixed_pages: number
  scanned_pages: number
  empty_pages: number
  ocr_required_pages: number
  chapter_count: number
  confirmed_chapters: number
}
